File: python/queue-worker/worker.py
import os
import threading
import time

from dbos import DBOS, DBOSConfig, Queue

# Define constants and models
WF_PROGRESS_KEY = "workflow_progress"


# Define a queue on which the web server
# can submit workflows for execution.
Queue("workflow-queue")


# This background workflow is submitted by the
# web server. It runs a number of steps,
# periodically reporting its progress.
@DBOS.workflow()
def workflow(num_steps: int):
    progress = {
        "steps_completed": 0,
        "num_steps": num_steps,
    }
    # The server can query this event to obtain
    # the current progress of the workflow
    DBOS.set_event(WF_PROGRESS_KEY, progress)
    for i in range(num_steps):
        step(i)
        # Update workflow progress each time a step completes
        progress["steps_completed"] = i + 1
        DBOS.set_event(WF_PROGRESS_KEY, progress)


@DBOS.step()
def step(i: int):
    print(f"Step {i} completed!")
    time.sleep(1)


# Configure and launch DBOS
if __name__ == "__main__":
    system_database_url = os.environ.get(
        "DBOS_SYSTEM_DATABASE_URL", "sqlite:///dbos_queue_worker.sqlite"
    )
    config: DBOSConfig = {
        "name": "dbos-queue-worker",
        "system_database_url": system_database_url,
    }
    DBOS(config=config)
    DBOS.launch()
    # After launching DBOS, the worker waits indefinitely,
    # dequeuing and executing workflows.
    threading.Event().wait()

File: typescript/queue-worker/worker.ts
import { DBOS, WorkflowQueue } from '@dbos-inc/dbos-sdk';

// Define constants
const WF_PROGRESS_KEY = 'workflow_progress';

// Define a queue on which the web server
// can submit workflows for execution.
new WorkflowQueue('workflow-queue');

// This background workflow is submitted by the
// web server. It runs a number of steps,
// periodically reporting its progress.
async function workflowFunction(numSteps: number): Promise<void> {
  const progress = {
    steps_completed: 0,
    num_steps: numSteps,
  };
  // The server can query this event to obtain
  // the current progress of the workflow
  await DBOS.setEvent(WF_PROGRESS_KEY, progress);
  for (let i = 0; i < numSteps; i++) {
    await DBOS.runStep(() => stepFunction(i), { name: `step-${i}` });
    // Update workflow progress each time a step completes
    progress.steps_completed = i + 1;
    await DBOS.setEvent(WF_PROGRESS_KEY, progress);
  }
}

export const workflow = DBOS.registerWorkflow(workflowFunction, {
  name: 'workflow',
});

async function stepFunction(i: number): Promise<void> {
  console.log(`Step ${i} completed!`);
  await new Promise((resolve) => setTimeout(resolve, 1000));
}

// Configure and launch DBOS
async function main(): Promise<void> {
  const systemDatabaseUrl =
    process.env.DBOS_SYSTEM_DATABASE_URL || 'postgresql://postgres:dbos@localhost:5432/dbos_queue_worker';
  DBOS.setConfig({
    name: 'dbos-queue-worker',
    systemDatabaseUrl: systemDatabaseUrl,
  });
  await DBOS.launch();
  // After launching DBOS, the worker waits indefinitely,
  // dequeuing and executing workflows.
  console.log('Worker started, waiting for workflows...');
  await new Promise(() => {});
}

main().catch(console.log);
